tests/extension/oneapi_device_global/device_global_functional_several_kernels_one_device.cpp

```diff
@@ -0,0 +1,154 @@
+/*******************************************************************************
+//
+//  SYCL 2020 Conformance Test Suite
+//
+//  Provides functional tests for device_global
+//
+//  Run several kernels on one device. The test modify device_global value in
+//  the first_kernel and then expect to read the same value from the second
+//  kernel. The test have a result array, that kernel uses through a buffer
+//  accessor. For the result array defined enum class is defined that have to be
+//  used for addressing elements in the array.
+//
+*******************************************************************************/
+
+#include "../../common/common.h"
+#include "../../common/type_coverage.h"
+#include "../../util/array.h"
+#include "device_global_common.h"
+#include "type_pack.h"
+
+#define TEST_NAME device_global_functional_several_kernel_in_one_device
+
+namespace TEST_NAMESPACE {
+using namespace sycl_cts;
+using namespace device_global_common_functions;
+#if defined(SYCL_EXT_ONEAPI_PROPERTY_LIST) && \
+    defined(SYCL_EXT_ONEAPI_DEVICE_GLOBAL)
+namespace oneapi = sycl::ext::oneapi;
+
+namespace several_kernel_in_one_device {
+template <typename T>
+oneapi::device_global<T> dev_global;
+template <typename T>
+oneapi::device_global<T> const_dev_global;
+
+template <typename T>
+struct first_kernel;
+template <typename T>
+struct second_kernel;
+
+// Used to address elements in the result array
+enum class indx : size_t {
+  const_expected,
+  non_const_expected,
+  size  // must be last
+};
+constexpr auto integral(const indx& i) { return to_integral<indx>(i); }
+
+/**
+ * @brief The function tests that device_global value changes correctly being
+ * modified and read from the different kernels on one device
+ * @tparam T Type of underlying device_global value
+ */
+template <typename T>
+void run_test(util::logger& log, const std::string& type_name) {
+  // Changing non-const value in first kernel
+  auto queue = util::get_cts_object::queue();
+  {
+    queue.submit([&](sycl::handler& cgh) {
+      cgh.single_task<first_kernel<T>>([=](sycl::kernel_handler h) {
+        value_helper<T>::change_val(dev_global<T>);
+      });
+    });
+    queue.wait_and_throw();
+  }
+
+  // Using remove_extent for get type of element when T is array
+  // If T is not array then element_type == T
+  // As we fill the array with the same numbers with type of the array element,
+  // we can store only one number to compare whole array with this number
+  using element_type = typename std::remove_extent_t<T>;
```

Comment:
Do we need the `typename` here?

Reply:
Good catch. We actually don't need it. I removed the `typename` here 3f512947af398d2ca0aa730c7c3b11276e411931